src/audio/codec_adapter/codec/waves.c

```diff
@@ -0,0 +1,726 @@
+	MaxxBuffer_t            i_buffer;
+	MaxxBuffer_t            o_buffer;
+	uint32_t                response_max_bytes;
+	uint32_t                request_max_bytes;
+	void                    *response;
+};
+
+enum waves_codec_params {
+	PARAM_NOP = 0,
+	PARAM_MESSAGE = 1,
+	PARAM_REVISION = 2
+};
+
+static int32_t sample_format_convert_to_bytes(MaxxBuffer_Format_t fmt)
+{
+	/* converts MaxxBuffer_Format_t to number of bytes it requires
+	 * -EINVAL if format not supported
+	 */
+	int32_t res;
+
+	switch (fmt) {
+	case MAXX_BUFFER_FORMAT_Q1_15:
+		res = 2;
+		break;
+	case MAXX_BUFFER_FORMAT_Q1_23:
+		res = 3;
+		break;
+	case MAXX_BUFFER_FORMAT_Q9_23:
+	case MAXX_BUFFER_FORMAT_Q1_31:
+	case MAXX_BUFFER_FORMAT_FLOAT:
+	case MAXX_BUFFER_FORMAT_Q5_27:
+		res = 4;
+		break;
+	default:
+		res = -EINVAL;
+		break;
+	}
+	return res;
+}
+
+static MaxxBuffer_Format_t sample_format_convert_sof_to_me(enum sof_ipc_frame fmt)
+{
+	/* converts enum sof_ipc_frame to MaxxBuffer_Format_t
+	 * -EINVAL if format not supported
+	 */
+	MaxxBuffer_Format_t res;
+
+	switch (fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		res = MAXX_BUFFER_FORMAT_Q1_15;
+		break;
+	case SOF_IPC_FRAME_S24_4LE:
+		res = MAXX_BUFFER_FORMAT_Q9_23;
+		break;
+	case SOF_IPC_FRAME_S32_LE:
+		res = MAXX_BUFFER_FORMAT_Q1_31;
+		break;
+	case SOF_IPC_FRAME_FLOAT:
+		res = MAXX_BUFFER_FORMAT_FLOAT;
+		break;
+	default:
+		res = -EINVAL;
+		break;
+	}
+	return res;
+}
+
+static MaxxBuffer_Layout_t buffer_format_convert_sof_to_me(uint32_t fmt)
+{
+	/* converts sof frame format to MaxxBuffer_Layout_t
+	 * -EINVAL if format not supported
+	 */
+	MaxxBuffer_Layout_t res;
+
+	switch (fmt) {
+	case SOF_IPC_BUFFER_INTERLEAVED:
+		res = MAXX_BUFFER_LAYOUT_INTERLEAVED;
+		break;
+	case SOF_IPC_BUFFER_NONINTERLEAVED:
+		res = MAXX_BUFFER_LAYOUT_DEINTERLEAVED;
+		break;
+	default:
+		res = -EINVAL;
+		break;
+	}
+	return res;
+}
+
+static void trace_array(const struct comp_dev *dev, const uint32_t *arr, uint32_t len)
+{
+	/* traces uint32_t array of len elements into debug messages */
+	uint32_t i;
+
+	for (i = 0; i < len; i++)
+		comp_dbg(dev, "trace_array() data[%03d]:0x%08x", i, *(arr + i));
+}
+
+static void waves_codec_data_clear(struct waves_codec_data *waves_codec)
+{
+	/* initiallizes waves_codec_data to default values */
+	int i;
+	uint8_t *byte = (uint8_t *) waves_codec;
+
+	for (i = 0; i < sizeof(struct waves_codec_data); i++)
+		*byte++ = 0;
+}
+
+static int waves_effect_check(struct comp_dev *dev)
+{
+	/* checks if stream parameters fit MaxxEffect
+	 * returnes non-zero value in case something is not supported
+	 * can be extended to utilize MaxxEffect scenarios for future products
+	 */
+	struct comp_data *component = comp_get_drvdata(dev);
+	const struct audio_stream *src_fmt = &component->ca_source->stream;
+	const struct audio_stream *snk_fmt = &component->ca_sink->stream;
+
+	comp_dbg(dev, "waves_effect_check() start");
+
+	/* resampling not supported */
+	if (src_fmt->rate != snk_fmt->rate) {
```

Comment:
ack, IIRC - I think we have a component/buffer API call for this now. e.g. do rates, formats, channels match ?

Reply:
hi @mrajwa, @lgirdwood 

Is it comp_verify_params, or something else?

Reply:
Hi @mrajwa,

Is there a way to ask specific source/sink format from codec_adapter or make codec_adapter match source and sink?
I am working on passing alsa conformance test on chromebook and see that some tests are doing source and sink with different sample formats. 
What is a correct way to handle it in codec_adapter?

Reply:
@lgirdwood
I see `comp_verify_params` is being called on codec_adapter level in `codec_adapter_params` function, which is not present on codec level.
What is recommended way to do stream/component alignment verification? 
Should custom codec somehow report requirements to codec_adapter and then codec_adapter will validate them during `codec_adapter_params` call?

Reply:
Hi @lgirdwood, @cujomalainey, @mrajwa,

Please suggest how to use `comp_verify_params` to validate stream from codec level?

Reply:
Hmm, this seems like something missing in the codec adapter api, I feel a lot could be simplified if the adapter could define the flags passed into that call. Ideally yes i think this should be called once and by the adapter, in the mean time though we need to figure out if calling it twice will have any ill affects, otherwise we will have to continue with the current method.

Reply:
@cujomalainey ok, I will check it. Will current implementation be accepted?

Reply:
yes, just please add a TODO for us to track here, something like "use fallback to comp_verify_params when ready"

Reply:
ok, done